src/clr-angular/data/datagrid/render/row-renderer.spec.ts

```diff
@@ -3,94 +3,51 @@
-import { Items } from '../providers/items';
-import { Page } from '../providers/page';
-import { RowActionService } from '../providers/row-action-service';
-import { Selection } from '../providers/selection';
-import { Sort } from '../providers/sort';
-import { StateDebouncer } from '../providers/state-debouncer.provider';
-
-import { DatagridCellRenderer } from './cell-renderer';
-import { DomAdapter } from '../../../utils/dom-adapter/dom-adapter';
-import { DatagridRenderOrganizer } from './render-organizer';
-import { MOCK_ORGANIZER_PROVIDER, MockDatagridRenderOrganizer } from './render-organizer.mock';
+import { Component, DebugElement } from '@angular/core';
+import { DATAGRID_SPEC_PROVIDERS, TestContext } from '../helpers.spec';
 import { DatagridRowRenderer } from './row-renderer';
+import { ColumnsService } from '../providers/columns.service';
+import { DatagridColumnState } from '../interfaces/column-state.interface';
+import { By } from '@angular/platform-browser';
+import { DatagridCellRenderer } from './cell-renderer';
+import { BehaviorSubject } from 'rxjs';
 
-const PROVIDERS = [
-  Selection,
-  Items,
-  FiltersProvider,
-  Sort,
-  Page,
-  RowActionService,
-  ExpandableRowsCount,
-  MOCK_ORGANIZER_PROVIDER,
-  DomAdapter,
-  HideableColumnService,
-  DatagridWillyWonka,
-  StateDebouncer,
-  DisplayModeService,
-];
 export default function(): void {
+  let columnsService: ColumnsService;
+
+  function initService(cols = 2) {
+    for (let i = 0; i < cols; i++) {
+      columnsService.columns[i] = new BehaviorSubject<DatagridColumnState>({ width: i });
+    }
+  }
+
   describe('DatagridRowRenderer directive', function() {
     let context: TestContext<DatagridRowRenderer, SimpleTest>;
-    let organizer: MockDatagridRenderOrganizer;
-    let cellWidthSpy: jasmine.Spy;
+    let cells: DebugElement[];
 
     beforeEach(function() {
-      context = this.create(DatagridRowRenderer, SimpleTest, PROVIDERS);
-      organizer = <MockDatagridRenderOrganizer>context.getClarityProvider(DatagridRenderOrganizer);
-      organizer.widths = [{ px: 42, strict: false }, { px: 24, strict: true }];
-      cellWidthSpy = spyOn(DatagridCellRenderer.prototype, 'setWidth');
+      context = this.create(DatagridRowRenderer, SimpleTest, DATAGRID_SPEC_PROVIDERS);
+      columnsService = context.getClarityProvider(ColumnsService);
+      initService();
     });
 
-    it('sets the widths of the cells when notified', function() {
-      organizer.updateRenderStep.next(DatagridRenderStep.ALIGN_COLUMNS);
-      expect(cellWidthSpy.calls.allArgs()).toEqual([[false, 42], [true, 24]]);
-    });
-
-    it("doesn't set the width when the organizer doesn't have them yet", function() {
-      organizer.widths = [];
-      organizer.updateRenderStep.next(DatagridRenderStep.ALIGN_COLUMNS);
-      expect(cellWidthSpy).not.toHaveBeenCalled();
-    });
-
-    it('sets the widths of the cells when created after the widths have been computed', function() {
```

Comment:
Losing this test seems like a pretty big deal to me. It's important that rows added _after_ the width have been computed correctly render with the right width. It's the basis of how our rendering works with pagination, for instance.

To be honest, right now I'm not even sure this works in the current state of this refactor. I do not see how moving to the next page preserves the computed columns width and renders them on the new rows.

Reply:
I've put a test back in for something similar.